server/etcdserver/server.go

```diff
@@ -1223,6 +1223,15 @@ func (s *EtcdServer) TransferLeadership() error {
 		return nil
 	}
 
+	if active, safe := s.isQuorumActiveAfterOneMemberIsDown(0 /* healthInterval */); !safe {
+		lg.Info(
+			"skipped leadership transfer; local member is shutting down; leadership transfer requires active quorum",
```

Comment:
I would argue that leadership transfer still makes sense to transfer leadership as long there is a quorum before leader shutting down. If not it would be like a leader bailing and not telling anyone they are resigning. 

There is one detail I would like to point to: difference between passing on leadership vs informing of resignation. From code I just read it seems raft implements only first case. Resigning leader picks their successor, problem is that this requires quorum. In second case leader would just inform followers that they are resigning and they can start election early. This eliminates "surprise" after leader disappears.

Question: Why are you proposing to skip leader transfer when there is still quorum?  Difference between `isQuorumActiveAfterOneMemberIsDown` vs `isQuorumActive`.

Reply:
cc @tbg

Reply:
@serathius the motivation is in the 3-member test cluster termination, one member is down, leader is unnecessarily waiting for 7s to wait for leadership transfer response.. 

```
/Users/wachao/go/src/github.com/ahrtr/etcd/bin/etcd (TestMemberAddNotLearnerDisableStrictReconfigCheckNoWaitForQuorumPeerTLS-test-1) (85679): 
{
    "level": "info",
    "ts": "2023-04-12T13:21:47.724134+0800",
    "caller": "etcdserver/server.go:1178",
    "msg": "leadership transfer starting",
    "local-member-id": "d70245c72180528a",
    "current-leader-member-id": "d70245c72180528a",
    "transferee-member-id": "c61f33e98087dec2"
}
...
/Users/wachao/go/src/github.com/ahrtr/etcd/bin/etcd (TestMemberAddNotLearnerDisableStrictReconfigCheckNoWaitForQuorumPeerTLS-test-1) (85679): 
{
    "level": "warn",
    "ts": "2023-04-12T13:21:54.724839+0800",
    "caller": "etcdserver/server.go:1257",
    "msg": "leadership transfer failed",
    "local-member-id": "d70245c72180528a",
    "error": "etcdserver: request timed out, leader transfer took too long"
}
```

As far as I know, there is no resignation version of leadership transfer. It can be done outside of raft like this PR.

Reply:
Yeah, you are right.

Reply:
> So instead of checking quorum could we set a shorter timeout and do repeated checks to longestConnected? It will fail is there is no followers to pass leadership to.

Sounds good to me.

Reply:
> So the problem is not shutting down when there is no quorum, problem is when leaderTransfer fails as followers are shutdown at the same time.

Agreed.

> So instead of checking quorum could we set a shorter timeout and do repeated checks to longestConnected?

Why do repeated checks? Either way (repeated check, e.g. 3 times, or transfer leadership to a member being shutdown concurrently), there is a timeout. Instead of complicating the implementation, It might be better just keep it as it's. Reasons:
1. Usually in production env, it's rare for multiple multiple members being shutdown concurrently;
2. Even it happens, raft can automatically elect a new leader;